timemachine/gcmc/mover.py

```diff
@@ -0,0 +1,259 @@
+        #   a. if turning off, compute delta_U when non-interacting
+        #   b. if turning on, randomly rotate and translate, then turn on and compute delta_Us
+
+        # print("num waters", np.sum(water_states))
+
+        # TBD: should we compute cur_params from water_states?
+
+        n_waters = len(water_states)
+        n_atoms = len(coords)
+
+        chosen_water = np.random.randint(0, n_waters)
+        chosen_water_atoms = water_idxs[chosen_water]
+
+        a_idxs = chosen_water_atoms
+        b_idxs = np.setdiff1d(np.arange(n_atoms), chosen_water_atoms)
+
+        self.t_count += 1
+
+        box_vol = np.product(np.diag(box))
+
+        beta = 1 / DEFAULT_KT
+        adams_B = beta * CHEMICAL_POTENTIAL + np.log(box_vol / STANDARD_VOLUME)
+
+        # useful for visualization
+        TRANSLATIONAL_OFFSET = 0
+
+        if water_states[chosen_water]:
+            # turn interacting water off
+
+            trial_params = cur_params.copy()
+
+            np.testing.assert_equal(trial_params[chosen_water_atoms], self.on_params)
+
+            # compute acceptance probability, the - sign is because we're missing
+            # these interactions
+            delta_U = -self.U_fn(coords, box, trial_params, a_idxs, b_idxs)
+            # print("delta_U delete", delta_U)
+            # assert delta_U < 1e4
+
+            # p_delete = min(1, np.exp(-adams_B) * np.exp(-beta * delta_U))
+            p_delete = min(1, np.exp(-adams_B - beta * delta_U))
+            trial_params[chosen_water_atoms] = self.off_params
+
+            if p_delete > np.random.rand():
+                # move accepted
+                next_water_state = water_states.copy()
+                next_water_state[chosen_water] = 0
+                # print("deletion success")
+                self.d_count += 1
+                trial_coords = coords.copy()
+                trial_coords[chosen_water_atoms] += np.ones(3) * TRANSLATIONAL_OFFSET
+                return trial_coords, trial_params, next_water_state
+            else:
+                # move rejected
+                return coords, cur_params, water_states
+
+        else:
+            # turn non-interacting water on
+
+            trial_params = cur_params.copy()
+            trial_params[chosen_water_atoms] = self.on_params
+
+            # randomly rotate and translate the chosen water
+            trial_chosen_coords = coords[chosen_water_atoms]
+            trial_translation = np.diag(box) * np.random.rand(3) + TRANSLATIONAL_OFFSET
+
+            # tbd - velocities?
+            moved_coords = randomly_rotate_and_translate(trial_chosen_coords, trial_translation)
+            trial_coords = coords.copy()
+            trial_coords[chosen_water_atoms] = moved_coords
+
+            delta_U = self.U_fn(trial_coords, box, trial_params, a_idxs, b_idxs)
+
+            # p_insert = min(1, np.exp(adams_B) * np.exp(-beta * delta_U))
+            p_insert = min(1, np.exp(adams_B - beta * delta_U))
```

Comment:
Same point of uncertainty as in the deletion branch: I see why acceptance probability of this proposal should depend on chemical potential, but I'm not quite following why it depends on the standard state volume of water.

Clarity nit: consider renaming both `p_insert` and `p_delete` to `p_accept`.

Reply:
willfix clarity nit.